### PR TITLE
Render control plane from cluster chart

### PR DESCRIPTION
**Pull request by nprokopic**
Towards https://github.com/giantswarm/roadmap/issues/3047

### What this PR does / why we need it

Render control plane resources from cluster chart.

### Checklist

- [ ] Update changelog in CHANGELOG.md.

### Trigger e2e tests

/run cluster-test-suites

### CODE REVIEW

File: .circleci/ct-config.yaml [hunk 1]
```diff
@@ -1,3 +1,4 @@
 ---
 chart-repos:
   - cluster-catalog=https://giantswarm.github.io/cluster-catalog
+  - cluster-test-catalog=https://giantswarm.github.io/cluster-test-catalog
```
Comment (nprokopic) [inline .circleci/ct-config.yaml:4]: Needed for using a non-released version of cluster chart, will be reverted before merging.

File: .github/workflows/zz_generated.diff_helm_render_templates.yaml [hunk 2]
```diff
@@ -82,6 +82,7 @@ jobs:
           mkdir -p /tmp/${{ matrix.values }}
           # TODO split files by "API"_"KIND"_"NAMESPACE|clusterwide"_"NAME"
           helm repo add cluster-catalog https://giantswarm.github.io/cluster-catalog/
+          helm repo add cluster-test-catalog https://giantswarm.github.io/cluster-test-catalog/
```
Comment (nprokopic) [inline .github/workflows/zz_generated.diff_helm_render_templates.yaml:85]: Needed for using a non-released version of cluster chart, will be reverted before merging.

File: helm/cluster-aws/Chart.yaml [hunk 3]
```diff
@@ -19,5 +19,5 @@ dependencies:
     version: "0.6.5"
     repository: "https://giantswarm.github.io/cluster-catalog"
   - name: cluster
-    version: "0.1.2"
-    repository: "https://giantswarm.github.io/cluster-catalog"
+    version: "0.1.2-cfbe561614ceabf4016216e8b519d22a35875e4c"
```
Comment (nprokopic) [inline helm/cluster-aws/Chart.yaml:22]: This will be updated to most probably v0.2.0 before merging the PR.

File: helm/cluster-aws/templates/_control_plane.tpl [hunk 4]
```diff
@@ -48,6 +48,7 @@ subnet:
 {{- end }}
 
 {{- define "control-plane" }}
+{{- if not $.Values.cluster.providerIntegration.resourcesApi.controlPlaneResourceEnabled }}
```
Comment (nprokopic) [inline helm/cluster-aws/templates/_control_plane.tpl:51]: I will remove the entire KubeadmControlPlane template before merging the PR.

File: helm/cluster-aws/values.schema.json [hunk 5]
```diff
@@ -186,11 +186,153 @@
             "default": {
                 "providerIntegration": {
                     "clusterAnnotationsTemplateName": "awsConnectivityLabels",
+                    "components": {
+                        "systemd": {
+                            "timesyncd": {
+                                "ntp": [
+                                    "169.254.169.123"
```
Comment (nprokopic) [inline helm/cluster-aws/values.schema.json:193]: Setting AWS-specific NTP server to Amazon Time Sync Service.

File: helm/cluster-aws/values.schema.json [hunk 6]
```diff
@@ -186,11 +186,153 @@
...
+                                ]
+                            }
+                        }
+                    },
+                    "connectivity": {
+                        "proxy": {
+                            "noProxy": {
```
Comment (nprokopic) [inline helm/cluster-aws/values.schema.json:200]: AWS-specific configuration of noProxy:
- 2 fixed addresses:
  - AWS ELB domain: `elb.amazonaws.com`
  - AWS IMDS: `169.254.169.254`
- More noProxy addresses from the template (where Helm templating can be used to customize the valeus in an array, something we can't do in plain JSON here)
  - The template is currently rendering `$.Values.global.connectivity.network.vpcCidr`

Other provider-independent values are added in the cluster chart in `cluster.internal.kubeadm.proxy.noProxyList` Helm template here https://github.com/giantswarm/cluster/blob/1ed268a2aab96a9dddf48206952935e60ebaa0e3/helm/cluster/templates/clusterapi/_helpers.tpl#L1.

File: helm/cluster-aws/values.schema.json [hunk 7]
```diff
@@ -186,11 +186,153 @@
...
+                                "addressesTemplate": "awsNoProxyList"
+                            }
+                        }
+                    },
                     "controlPlane": {
+                        "kubeadmConfig": {
+                            "clusterConfiguration": {
+                                "apiServer": {
+                                    "additionalAdmissionPlugins": [
+                                        "PodSecurityPolicy"
```
Comment (nprokopic) [inline helm/cluster-aws/values.schema.json:214]: We can maybe drop this, if https://github.com/giantswarm/cluster-aws/pull/481 gets merged first 🎉

File: helm/cluster-aws/values.schema.json [hunk 8]
```diff
@@ -186,11 +186,153 @@
...
+                                    ],
+                                    "apiAudiences": {
+                                        "templateName": "awsApiServerApiAudiences"
```
Comment (nprokopic) [inline helm/cluster-aws/values.schema.json:217]: A name of a Helm template that is reading API server `api-audiences` value. The value is calculated (based on the region name from template `"aws-region"`), so it cannot be set in JSON.

File: helm/cluster-aws/values.schema.json [hunk 9]
```diff
@@ -186,11 +186,153 @@
...
+                                    },
+                                    "featureGates": [
+                                        {
+                                            "enabled": true,
+                                            "name": "CronJobTimeZone"
+                                        }
+                                    ],
+                                    "serviceAccountIssuer": {
+                                        "clusterDomainPrefix": "irsa"
```
Comment (nprokopic) [inline helm/cluster-aws/values.schema.json:226]: For now we have two options to set API server `service-account-issuer`:
1. The one used here where cluster-<provider> app sets the prefix in the Helm values, and then this is prepended to `<cluster name>.<base domain>` (like now in cluster-aws). See cluster chart code here https://github.com/giantswarm/cluster/blob/1ed268a2aab96a9dddf48206952935e60ebaa0e3/helm/cluster/templates/clusterapi/controlplane/_helpers_clusterconfiguration_apiserver.tpl#L84.
2. Just set the whole URL with `serviceAccountIssuer.url` instead of setting prefix in `serviceAccountIssuer.clusterDomainPrefix`.

File: helm/cluster-aws/values.schema.json [hunk 10]
```diff
@@ -186,11 +186,153 @@
...
+                                    }
+                                }
+                            },
+                            "ignition": {
+                                "containerLinuxConfig": {
+                                    "additionalConfig": {
+                                        "storage": {
+                                            "filesystems": [
```
Comment (nprokopic) [inline helm/cluster-aws/values.schema.json:234]: AWS-specific Ignition config for filesystems, while e.g. cluster-azure ATM has a different solution. IIUC both are workarounds needed due to an upstream bug https://github.com/kubernetes-sigs/cluster-api/issues/7679.

For now we keep provider-specific config here (so for AWS cluster chart will render the same template like cluster-aws now), and later we will try to have a unified solution in the cluster chart (either proper usage after upstream bug fix, or an aligned workaround if possible).

With a unified solution in the cluster chart, the provider-specific part of the config should ideally boil down to just specifying provider-specific values like device.

File: helm/cluster-aws/values.schema.json [hunk 11]
```diff
@@ -186,11 +186,153 @@
...
+                                                        "format": "xfs",
+                                                        "label": "kubelet",
+                                                        "wipeFilesystem": true
+                                                    },
+                                                    "name": "kubelet"
+                                                }
+                                            ]
+                                        },
+                                        "systemd": {
+                                            "units": [
```
Comment (nprokopic) [inline helm/cluster-aws/values.schema.json:265]: Provider-independent systemd units are all now in the cluster chart, except these 3 mount units here. For these remaining mount units see the above comment about filesystems.

File: helm/cluster-aws/values.schema.json (hunk 6 above)
Comment (njuettner) [reply helm/cluster-aws/values.schema.json:200]: There's something fishy with the noProxy list

Templating adds an `,<no value>`
```
--
      noProxy: 127.0.0.1,localhost,svc,local,169.254.169.254,10.0.0.0/16,172.31.0.0/16,100.64.0.0/12,test-mc-proxy.example.com,elb.amazonaws.com,<no
        value>
```

Comment (nprokopic) [reply helm/cluster-aws/values.schema.json:200]: Hmm, fishy yeah, thanks! Checking this.

File: helm/cluster-aws/values.schema.json (hunk 8 above)
Comment (nprokopic) [reply helm/cluster-aws/values.schema.json:217]: Right, good point. I guess just `template` should be clear enough.

File: helm/cluster-aws/values.schema.json (hunk 6 above)
Comment (nprokopic) [reply helm/cluster-aws/values.schema.json:200]: Where did you see that this `,<no value>` is added?

When I render templates on my machine with these CI values https://github.com/giantswarm/cluster-aws/blob/master/helm/cluster-aws/ci/test-mc-proxy.yaml (**noProxy is not specified in Helm values**):

Without cluster chart I get a dangling comma in the end of noProxy string, e.g. proxy conf file looks like this:

```
[Service]
Environment="HTTP_PROXY=http://proxy.mcproxy.example.com:4000"
Environment="HTTPS_PROXY=http://proxy.mcproxy.example.com:4000"
Environment="NO_PROXY=127.0.0.1,localhost,svc,local,169.254.169.254,10.0.0.0/16,172.31.0.0/16,100.64.0.0/12,test-mc-proxy.example.com,elb.amazonaws.com,"
Environment="http_proxy=http://proxy.mcproxy.example.com:4000"
Environment="https_proxy=http://proxy.mcproxy.example.com:4000"
Environment="no_proxy=127.0.0.1,localhost,svc,local,169.254.169.254,10.0.0.0/16,172.31.0.0/16,100.64.0.0/12,test-mc-proxy.example.com,elb.amazonaws.com,"
```

With cluster chart there is no dangling comma:

```
[Service]
Environment="HTTP_PROXY=http://proxy.mcproxy.example.com:4000"
Environment="HTTPS_PROXY=http://proxy.mcproxy.example.com:4000"
Environment="NO_PROXY=127.0.0.1,localhost,svc,local,test-mc-proxy.example.com,172.31.0.0/16,100.64.0.0/12,elb.amazonaws.com,169.254.169.254,10.0.0.0/16"
Environment="http_proxy=http://proxy.mcproxy.example.com:4000"
Environment="https_proxy=http://proxy.mcproxy.example.com:4000"
Environment="no_proxy=127.0.0.1,localhost,svc,local,test-mc-proxy.example.com,172.31.0.0/16,100.64.0.0/12,elb.amazonaws.com,169.254.169.254,10.0.0.0/16"
```

When noProxy is specified in Helm values then rendering looks fine when rendered from both cluster-aws and cluster.

Comment (njuettner) [reply helm/cluster-aws/values.schema.json:200]: helm template --dry-run --debug -f ci/test-mc-proxy.yaml  --generate-name .|grep noProxy

Comment (nprokopic) [reply helm/cluster-aws/values.schema.json:200]: New cluster chart version (yet to be released) has a template that can be used also in correctly renders the NO_PROXY values.

File: helm/cluster-aws/values.schema.json (hunk 8 above)
Comment (nprokopic) [reply helm/cluster-aws/values.schema.json:217]: Changes implemented in cluster chart, I'll update this PR as soon as we release new version of the cluster chart.

Comment (nprokopic) [reply helm/cluster-aws/values.schema.json:217]: PR updated to use the latest cluster chart, so these Helm value names are now aligned.

Comment (nprokopic) [reply helm/cluster-aws/values.schema.json:217]: Closing this conversation to reduce the noise a bit here in the PR. Fee free to re-open if needed.

File: helm/cluster-aws/values.schema.json (hunk 6 above)
Comment (nprokopic) [reply helm/cluster-aws/values.schema.json:200]: PR updated with new cluster chart, so this PR now uses noProxy template from cluster chart.

Closing this conversation to reduce the noise a bit here in the PR. Fee free to re-open if needed.

File: helm/cluster-aws/templates/_control_plane.tpl (hunk 4 above)
Comment (mnitchev) [reply helm/cluster-aws/templates/_control_plane.tpl:51]: Is this still the case or are we keeping the enable flag?

File: helm/cluster-aws/values.schema.json (hunk 9 above)
Comment (mnitchev) [reply helm/cluster-aws/values.schema.json:226]: I think sticking with option 1 is fine

**Review by mnitchev — APPROVED** [review]
LGTM just the flag comment.

File: helm/cluster-aws/templates/_control_plane.tpl (hunk 4 above)
Comment (nprokopic) [reply helm/cluster-aws/templates/_control_plane.tpl:51]: Still the case, I will remove the old code before merging it. Just kept it so I more easily compare before vs after Helm rendering while making changes locally.

P.S. We have the CI for the Helm rendering diff here in PRs, but it's a PITA to compare diff in files, because they get base64 encoded, so I do it more quickly and more reliable on my machine.

File: helm/cluster-aws/values.schema.json (hunk 9 above)
Comment (calvix) [reply helm/cluster-aws/values.schema.json:226]: option 2 may be necessary for CHINA where there we need to use cloudfront

**Comment by calvix** [comment]
I am just a little bit concerned about how vastly is the `values.scheme.json` expanding due to the need to propagate some values we had as files or templates before. It will be very hard for the customer to read the file and not to fucked up something by overriding some important default value or to figure out what they need.

**Comment by nprokopic** [comment]
> I am just a little bit concerned about how vastly is the values.scheme.json expanding due to the need to propagate some values we had as files or templates before. It will be very hard for the customer to read the file and not to fucked up something by overriding some important default value or to figure out what they need.

Absolutely legit comment, I have same concerns. It will definitely not stay like that.

There are some improvements that I would like to do (just when we're done with main part of the restructuring work).

TL;DR make schema smaller, and improve docs so you don't even have to read the schema.

More specifically:
- Move some re-usable generic parts of the schema to https://github.com/giantswarm/schema, so values.schema.json is smaller. e.g. we don't rally need a detailed systemd units schema in cluster-aws or in cluster chart, we can just have some reference like `"$ref": "https://schema.giantswarm.io/systemd/unit/v0.1.0"`.
- Improve rendering of schema docs, because currently it's not very well readable.
- As an extension of the above, write custom (not generated) guideline docs, so if you want to learn how customise something, you don't dig into JSON schema, but you just read some docs about it.
- Further improve validations in the schema, so it's hard to fuck up something.

**Review by calvix — APPROVED** [review]

**Comment by nprokopic** [comment]
To shed some light on files diff, as the files are encoded, so you don't see the diff here.

KubeletConfiguration diff is there because we updated TLS cipher suites to included all recommended ones (recommended in k8s docs):

image

In proxy conf the order of NO_PROXY addresses changed and there is no dangling comma anymore:

Before:
```conf
[Service]
Environment="HTTP_PROXY=http://proxy.mcproxy.example.com:4000"
Environment="HTTPS_PROXY=http://proxy.mcproxy.example.com:4000"
Environment="NO_PROXY=127.0.0.1,localhost,svc,local,169.254.169.254,10.0.0.0/16,172.31.0.0/16,100.64.0.0/12,test-mc-proxy.example.com,elb.amazonaws.com,"
Environment="http_proxy=http://proxy.mcproxy.example.com:4000"
Environment="https_proxy=http://proxy.mcproxy.example.com:4000"
Environment="no_proxy=127.0.0.1,localhost,svc,local,169.254.169.254,10.0.0.0/16,172.31.0.0/16,100.64.0.0/12,test-mc-proxy.example.com,elb.amazonaws.com,"
```

After:

```conf
[Service]
Environment="HTTP_PROXY=http://proxy.mcproxy.example.com:4000"
Environment="HTTPS_PROXY=http://proxy.mcproxy.example.com:4000"
Environment="NO_PROXY=127.0.0.1,localhost,svc,local,test-mc-proxy.example.com,172.31.0.0/16,100.64.0.0/12,elb.amazonaws.com,169.254.169.254,10.0.0.0/16"
Environment="http_proxy=http://proxy.mcproxy.example.com:4000"
Environment="https_proxy=http://proxy.mcproxy.example.com:4000"
Environment="no_proxy=127.0.0.1,localhost,svc,local,test-mc-proxy.example.com,172.31.0.0/16,100.64.0.0/12,elb.amazonaws.com,169.254.169.254,10.0.0.0/16"
```

**Comment by nprokopic** [comment]
That last e2e test run does not look good 😕 will try again

/run cluster-test-suites

**Comment by nprokopic** [comment]
Re-running tests to debug issues

/run cluster-test-suites

**Comment by nprokopic** [comment]
Retrying after fixing a bug from the previous e2e test failures :crossed_fingers: 

/run cluster-test-suites

**Comment by nprokopic** [comment]
Hello world test should be fixed, it passed here when cert-manager got updated in default-apps-aws https://github.com/giantswarm/default-apps-aws/pull/395

Retrying once again, to see if the failure is a flake, and to check more details about the failed certs I saw and commented [here](https://github.com/giantswarm/roadmap/issues/3151#issuecomment-1911921349).

/run cluster-test-suites

**Comment by nprokopic** [comment]
Retrying after e2e test changes (increased min nodes to 2) 🤞 

/run cluster-test-suites

**Comment by nprokopic** [comment]
✅ Private Suite passes

✅ Upgrade Suite passed

Standard test is failing with this:

```
[FAILED] Timed out after 600.003s.
  The function passed to Eventually returned the following error:
      <*url.Error | 0xc000d3b8f0>: 
      Get "https://hello-world.t-tlyhimt6zplvpicasv.gaws.gigantic.io/": dial tcp: lookup [hello-world.t-tlyhimt6zplvpicasv.gaws.gigantic.io](http://hello-world.t-tlyhimt6zplvpicasv.gaws.gigantic.io/) on 172.31.0.10:53: server misbehaving
      {
          Op: "Get",
          URL: "https://hello-world.t-tlyhimt6zplvpicasv.gaws.gigantic.io/",
          Err: <*net.OpError | 0xc000016d20>{
              Op: "dial",
              Net: "tcp",
              Source: nil,
              Addr: nil,
              Err: <*net.DNSError | 0xc0005eeb80>{
                  Err: "server misbehaving",
                  Name: "[hello-world.t-tlyhimt6zplvpicasv.gaws.gigantic.io](http://hello-world.t-tlyhimt6zplvpicasv.gaws.gigantic.io/)",
                  Server: "172.31.0.10:53",
                  IsTimeout: false,
                  IsTemporary: true,
                  IsNotFound: false,
              },
          },
      }
  In [It] at: /app/internal/common/hello.go:170 @ 01/30/24 13:13:43.278
```

File: helm/cluster-aws/templates/_control_plane.tpl [hunk 12]
```diff
@@ -48,208 +48,6 @@ subnet:
 {{- end }}
 
 {{- define "control-plane" }}
-apiVersion: controlplane.cluster.x-k8s.io/v1beta1
```
Comment (nprokopic) [inline helm/cluster-aws/templates/_control_plane.tpl:51]: This looks scary 😄

**Comment by nprokopic** [comment]
Another run after actually removing the KubeadmControlPlane resource

/run cluster-test-suites

**Comment by nprokopic** [comment]
Same result again:

✅ Private Suite passes

✅ Upgrade Suite passed, with Hello World test passing.

❌ Standard test is failing with what should be the known intermittent network issue with hello world test.

It's been a long one, merging this now!
